sdk/nodejs/policy/policy.ts

```diff
@@ -121,42 +121,29 @@ export interface ResourceValidationArgs {
 
 /**
  * A helper function that returns a strongly-typed resource validation function.
- * @param typeFilter A type guard used to determine if the args are an instance of the resource.
+ * @param resourceClass A resource class used to filter this check to only resources of the specified class and
+ * determine the appropriate strongly-typed `TArg` type to use for the resource.
  * @param validate A callback function that validates if the resource definition violates a policy.
  */
-export function validateTypedResource<TResource extends Resource>(
-    typeFilter: (o: any) => o is TResource,
+export function validateTypedResource<TResource extends Resource, TArgs>(
+    resourceClass: { new(name: string, args: TArgs, ...rest: any[]): TResource },
     validate: (
-        resource: q.ResolvedResource<TResource>,
+        props: NonNullable<Unwrap<TArgs>>,
         args: ResourceValidationArgs,
         reportViolation: ReportViolation) => Promise<void> | void,
 ): ResourceValidation {
     return (args: ResourceValidationArgs, reportViolation: ReportViolation) => {
-        args.props.__pulumiType = args.type;
-        if (typeFilter(args.props) === false) {
+        const isInstance = (<any>resourceClass).isInstance;
+        if (!isInstance || typeof isInstance !== "function") {
             return;
         }
-        validate(args.props as q.ResolvedResource<TResource>, args, reportViolation);
+        if (isInstance({ __pulumiType: args.type }) !== true) {
+            return;
+        }
+        validate(args.props as NonNullable<Unwrap<TArgs>>, args, reportViolation);
     };
 }
 
-/**
- * A helper function that returns `props` as a strongly-typed resolved resource based
- * on the specified `type` when `filter` returns true, otherwise `undefined` is returned.
- * @param typeFilter A type guard used to determine if the args are an instance of the resource.
- * @param args Argument bag for specifying the `type` and `props`.
- */
-export function asTypedResource<TResource extends Resource>(
```

Comment:
This helper is no longer generally useful. For `validateResource` we want the "input" shape, but for `validateStack`, we want the "output" shape.

It'd be best to simplify things here by removing this. For `validateResource`, if you want to write your validation function in a strongly-typed way, use `validateTypedResource`. For `validateStack`, if you want to filter/cast resources, use `@pulumi/query`. As a follow-up, I'll make sure we have an example of how to use `@pulumi/query` with `validateStack`.

Reply:
Definitely interested in seeing what this looks like!